### ISSUE TITLE
NVFuser should extend caching to remove necessity for PrimTorch's executor to Provide Tensor Contiguity Info

### ISSUE THREAD

**Issue by kevinstephano** [pytorch/pytorch#80551]
### 🚀 The feature, motivation and pitch

Currently, PrimTorch's executor provides contiguity and tensor rank size information in the `FusionDefinttion` for each fusion group sent to NVFuser.  This should be removed in favor of NVFuser receiving this information at runtime and caching the graph appropriately.

Therefore, the only information NVFuser should be provided in the Tensor Definition is the number of ranks of each Tensor. @mruberry does this match your understanding?

### Additional context

CC @csarofeen @jjsjann123 @ngimel @IvanYashchuk @mru

**Comment by IvanYashchuk** [comment]
@kevinstephano, I plan to add this to my caching PR https://github.com/pytorch/pytorch/pull/80525, it should be a very simple change there, only this line should be removed to use `define_tensor(ndims, type)` TensorView builder: https://github.com/pytorch/pytorch/blob/3a6d6bc3cc9bda07d398ea03e838e223e96717b5/torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp#L191

**Comment by IvanYashchuk** [comment]
One complication could be that currently `prims.broadcast_in_dim` records concrete/non-symbolic shapes during tracing.

**Comment by kevinstephano** [comment]
I don't think the question of contiguity caching is clear to me, yet.  So, I am hesitant to make changes along this path without more planning.

For instance, symbolic caching of the tensor rank sizes is happening below the Python Interface where we have multiple runtimes (kernels) per high level fusion definition based on the input tensor(s).  I am not sure we have plumbed that for contiguity.

**Comment by IvanYashchuk** [comment]
> For instance, symbolic caching of the tensor rank sizes is happening below the Python Interface where we have multiple runtimes (kernels) per high level fusion definition based on the input tensor(s). I am not sure we have plumbed that for contiguity.

I don't know the details, but using `define_tensor(size_t ndims, DataType dtype)` without setting the contiguity info (as suggested in https://github.com/pytorch/pytorch/issues/80551#issuecomment-1170929685) worked in my testing. Leaving the contiguity there as it is now produces incorrect results for tensors with the same number of dimensions but different noncontiguous strides.

**Comment by csarofeen** [comment]
That replay is part of @kevinstephano's plan.

**Comment by ezyang** [comment]
> One complication could be that currently prims.broadcast_in_dim records concrete/non-symbolic shapes during tracing.

Yes, I want to elaborate on this. If we ingest a graph where we don't know if a tensor actually is contiguous or not, it seems fine to defer this to runtime. However, if as part of trace acquisition or primtorch lowering we have already baked in a contiguity check (e.g., because user code branched on `is_contiguous`), there isn't much point in testing contiguity again, and in fact if you're at the point where input validity checking is a major performance cost, it is actively harmful to test for contiguity in nvfuser's cache, as dynamo *already* tested this condition when determining if its cached FX graph for the python frame is valid or not. As for as correctness is concerned it's immaterial if you throw out everything you know and retest it again at runtime, but given that we're in the business of making PT programs run faster I think it's important to take care with the design to account for the terminal, most optimized case.

**Comment by csarofeen** [comment]
We already validate tensors on the hot-path to make sure they match our vectorization scheme. I don't think we should invent latency issues until we hit them. nvFuser is a stand alone system that will validate and cache inputs sent into it, to make sure they match the kernels its generating.
